### PR TITLE
#9 extract

### LINKED ISSUE #9: Extract from Bluesky
# Description

This ticket creates an extract file that extracts the data from the data sources - Bluesky firehose

# Outcome

We are able to extract the data into a relevant data structure e.g. Pandas Dataframe or CSV file.

# User story

As a data analyst, I want to extract data from the Bluesky Firehose into a structured format like a Pandas DataFrame or CSV file, so that I can easily analyse and work with the data for insights.

### PR DESCRIPTION
**Pull request by keoghrmj**
Extraction for the Bluesky Firehose has been completed for now with the data being successfully extracted and uploaded to a csv file.

Closes #9

### CODE REVIEW

File: pipeline/extract.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+
+def format_text(text: str) -> str:
+    """Removes extra lines and whitespaces from text"""
+    # Remove consecutive empty lines
+    text = re.sub(r'\n\s*\n', '\n', text)
+
+    text = text.strip()
+    # Replace multiple spaces with a single space
+    text = re.sub(r'\s+', ' ', text)
```
Comment (SurinaCS) [inline pipeline/extract.py]: great use of regex here

File: pipeline/extract.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+def extract_text_from_bytes(raw: bytes) -> str:
+    """Extracts text from a raw Bluesky post"""
+    try:
+        json_data = json.dumps(raw, cls=JSONExtra, indent=2)
+        parsed_json = json.loads(json_data)
+        text = parsed_json.get('text')
+        return format_text(text)
+    except (TypeError, AttributeError):
+        return None
```
Comment (SurinaCS) [inline pipeline/extract.py]: could you add a specific message or make it a bit clearer what exactly is going to happen here?

File: pipeline/extract.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+
+def connect_and_write(topics: list[str]) -> None:
+    """Connect to BlueSky Firehose API and write data to CSV."""
+    ssl_context = ssl.create_default_context(cafile=certifi.where())
+
+    client = FirehoseSubscribeReposClient()
+    client.ssl_context = ssl_context
+
+    with open('output.csv', mode='w', newline='', encoding='utf-8') as csvfile:
```
Comment (SurinaCS) [inline pipeline/extract.py]: can you make output.csv into a global variable?

File: pipeline/extract.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        writer = csv.writer(csvfile, quotechar='"', quoting=csv.QUOTE_MINIMAL)
+        header = ['text', 'keyword']
```
Comment (SurinaCS) [inline pipeline/extract.py]: potentially another global?

File: pipeline/extract.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+        writer.writerow(header)
+
+        def start_firehose_extraction() -> None:
+            """Sarts the Bluesky firehose extraction"""
```
Comment (SurinaCS) [inline pipeline/extract.py]: great docstrings

File: pipeline/extract.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        json_data = json.dumps(raw, cls=JSONExtra, indent=2)
+        parsed_json = json.loads(json_data)
+        text = parsed_json.get('text')
+        return format_text(text)
+    except (TypeError, AttributeError):
+        return None
+
+
+def get_firehose_data(message: bytes, topics: list[str],
+                      csv_writer: csv.writer, csvfile: csv.writer) -> None:
```
Comment (SurinaCS) [inline pipeline/extract.py]: great typehints

**Review by SurinaCS — CHANGES_REQUESTED** [review]
great work, potentially extract some variables as globals

File: pipeline/bluesky_extract.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            if processed_post.py_type == "app.bsky.feed.post":
+                firehose_text = extract_text_from_bytes(raw_bytes)
+
+                for topic in topics:
+                    if topic in firehose_text:
+
+                        csv_writer.writerow([firehose_text, topic])
+                        csvfile.flush()
+                        print(f"Written: {firehose_text}")
+                        print("="*100)
```
Comment (RidwanHamid501) [inline pipeline/bluesky_extract.py:68]: Can we use logging here instead

File: pipeline/bluesky_extract.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+from atproto_client.models.utils import get_or_create
+from atproto_firehose import FirehoseSubscribeReposClient, parse_subscribe_repos_message
+
+
+class JSONExtra(json.JSONEncoder):
+    """Serializes raw objects (including CID-Content Identifier) as strings."""
+
+    def default(self, obj: bytes):
+        try:
+            return json.JSONEncoder.default(self, obj)
```
Comment (RidwanHamid501) [inline pipeline/bluesky_extract.py:19]: would we be able to do return super().default(obj) instead?

File: pipeline/bluesky_extract.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    for operation in repo_commit.ops:
+        if operation.action in ["create"] and operation.cid:
+            raw_bytes = car_file.blocks.get(operation.cid)
+            processed_post = get_or_create(raw_bytes, strict=False)
+
+            if processed_post.py_type == "app.bsky.feed.post":
+                firehose_text = extract_text_from_bytes(raw_bytes)
+
+                for topic in topics:
+                    if topic in firehose_text:
```
Comment (Kurt812) [inline pipeline/bluesky_extract.py:63]: Would it make sense to check for the topic not including when it's part of another word e.g. looking for spy and catching crispy? Or is this something for the transform?

File: pipeline/bluesky_extract.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+def extract_text_from_bytes(raw: bytes) -> str:
+    """Extracts text from a raw Bluesky post"""
+    try:
+        json_data = json.dumps(raw, cls=JSONExtra, indent=2)
+        parsed_json = json.loads(json_data)
+        text = parsed_json.get('text')
+        return format_text(text)
+    except (TypeError, AttributeError):
+        return None
```
Comment (RidwanHamid501) [inline pipeline/bluesky_extract.py:44]: could add some logging here

File: pipeline/bluesky_extract.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+def get_firehose_data(message: bytes, topics: list[str],
+                      csv_writer: csv.writer, csvfile: csv.writer) -> None:
+    """Handles incoming messages, parses data, and writes to a CSV file based on dynamic topics."""
+    repo_commit = parse_subscribe_repos_message(message)
+    if not isinstance(repo_commit, models.ComAtprotoSyncSubscribeRepos.Commit):
+        return
+    car_file = CAR.from_bytes(repo_commit.blocks)
+    for operation in repo_commit.ops:
+        if operation.action in ["create"] and operation.cid:
```
Comment (RidwanHamid501) [inline pipeline/bluesky_extract.py:55]: why in ["create"] instead of == "create"

File: pipeline/bluesky_extract.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+            raw_bytes = car_file.blocks.get(operation.cid)
+            processed_post = get_or_create(raw_bytes, strict=False)
+
+            if processed_post.py_type == "app.bsky.feed.post":
+                firehose_text = extract_text_from_bytes(raw_bytes)
```
Comment (RidwanHamid501) [inline pipeline/bluesky_extract.py:60]: What happens if firehose_text is empty - ''? 
Suggestion:
if not firehose_text: continue

File: pipeline/bluesky_extract.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    client.ssl_context = ssl_context
+
+    with open('output.csv', mode='w', newline='', encoding='utf-8') as csvfile:
+        writer = csv.writer(csvfile, quotechar='"', quoting=csv.QUOTE_MINIMAL)
+        header = ['text', 'keyword']
+
+        writer.writerow(header)
+
+        def start_firehose_extraction() -> None:
+            """Sarts the Bluesky firehose extraction"""
```
Comment (Kurt812) [inline pipeline/bluesky_extract.py:86]: Starts*

**Review by RidwanHamid501 — CHANGES_REQUESTED** [review]

File: pipeline/bluesky_extract.py (hunk 9 above)
Comment (keoghrmj) [reply pipeline/bluesky_extract.py]: That does make a lot of sense but maybe this is something that can be addressed in transform?

File: pipeline/bluesky_extract.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+
+def extract_text_from_bytes(raw: bytes) -> str:
+    """Extracts text from a raw Bluesky post"""
+    try:
+        json_data = json.dumps(raw, cls=JSONExtra, indent=2)
+        parsed_json = json.loads(json_data)
+        text = parsed_json.get('text')
+        return format_text(text)
+    except (TypeError, AttributeError) as e:
+        logging.error(f"Error extracting text: {e}")
```
Comment (SurinaCS) [inline pipeline/bluesky_extract.py:58]: great use of logging

File: pipeline/bluesky_extract.py [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+    ssl_context = ssl.create_default_context(cafile=certifi.where())
+
+    client = FirehoseSubscribeReposClient()
+    client.ssl_context = ssl_context
+
+    with open(CSV_OUTPUT_FILE, mode='w', newline='', encoding='utf-8') as csvfile:
+        writer = csv.writer(csvfile, quotechar='"', quoting=csv.QUOTE_MINIMAL)
+        logging.info(f"Created CSV output file: {CSV_OUTPUT_FILE}")
+
+        writer.writerow(HEADER)
```
Comment (SurinaCS) [inline pipeline/bluesky_extract.py:110]: good use of globals

**Review by SurinaCS — APPROVED** [review]
nice changes

**Review by RidwanHamid501 — APPROVED** [review]

**Pull request merged**